libs/ssh/server.go

```diff
@@ -0,0 +1,322 @@
+package ssh
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/databricks/cli/libs/cmdio"
+	"github.com/databricks/databricks-sdk-go"
+	"golang.org/x/sync/errgroup"
+)
+
+//go:embed jupyter-init.py
+var jupyterInitScript string
+
+// connectionsManager manages concurrent websocket clients and sends a shutdown signal if no
+// clients are connected for a specified duration.
+type connectionsManager struct {
+	maxClients      int
+	shutdownDelay   time.Duration
+	shutdownTimer   *time.Timer
+	shutdownTimerMu sync.Mutex
+	connections     map[string]*proxyConnection
+	connectionsMu   sync.Mutex
+	TimedOut        chan bool
+}
+
+func newConnectionsManager(maxClients int, shutdownDelay time.Duration) *connectionsManager {
+	cm := &connectionsManager{
+		maxClients:    maxClients,
+		shutdownDelay: shutdownDelay,
+		connections:   make(map[string]*proxyConnection),
+		TimedOut:      make(chan bool),
+	}
+	cm.startShutdownTimer()
+	return cm
+}
+
+func (cm *connectionsManager) TryAdd(id string, conn *proxyConnection) bool {
+	if len(cm.connections) >= cm.maxClients {
+		return false
+	}
+	cm.addConnection(id, conn)
+	cmdio.LogString(context.Background(), fmt.Sprintf("Client connected, current count: %d", len(cm.connections)))
+	cm.cancelShutdownTimer()
+	return true
+}
+
+func (cm *connectionsManager) addConnection(id string, conn *proxyConnection) {
+	cm.connectionsMu.Lock()
+	defer cm.connectionsMu.Unlock()
+	cm.connections[id] = conn
+}
+
+func (cm *connectionsManager) Get(id string) (*proxyConnection, bool) {
+	cm.connectionsMu.Lock()
+	defer cm.connectionsMu.Unlock()
+	conn, exists := cm.connections[id]
+	return conn, exists
+}
+
+func (cm *connectionsManager) Remove(id string) {
+	cm.removeConnection(id)
+	cmdio.LogString(context.Background(), fmt.Sprintf("Client disconnected, current count: %d", len(cm.connections)))
+	if len(cm.connections) <= 0 {
+		cm.startShutdownTimer()
+	}
+}
+
+func (cm *connectionsManager) removeConnection(id string) {
+	cm.connectionsMu.Lock()
+	defer cm.connectionsMu.Unlock()
+	delete(cm.connections, id)
+}
+
+func (cm *connectionsManager) startShutdownTimer() {
+	cm.shutdownTimerMu.Lock()
+	defer cm.shutdownTimerMu.Unlock()
+	if cm.shutdownTimer != nil {
+		cm.shutdownTimer.Stop()
+	}
+	cmdio.LogString(context.Background(), fmt.Sprintf("Starting shutdown timer for %v", cm.shutdownDelay))
```

Comment:
Same comment as above for logging. This should never use `context.Background()`.